ukvm/ukvm_hv_openbsd_x86_64.c

```diff
@@ -0,0 +1,254 @@
+        .vsi_ar = (sreg->type
+            | (sreg->s << 4)
+            | (sreg->dpl << 5)
+            | (sreg->p << 7)
+            | (sreg->l << 13)
+            | (sreg->db << 14)
+            | (sreg->g << 15)
+            | (sreg->unusable << X86_SREG_UNUSABLE_BIT)),
+        .vsi_base = sreg->base
+    };
+    return vsi;
+}
+
+void ukvm_hv_mem_size(size_t *mem_size) {
+    ukvm_x86_mem_size(mem_size);
+}
+
+void ukvm_hv_vcpu_init(struct ukvm_hv *hv, ukvm_gpa_t gpa_ep,
+        ukvm_gpa_t gpa_kend, char **cmdline)
+{
+    struct ukvm_hvb *hvb = hv->b;
+
+    struct vm_resetcpu_params vrp = {
+        .vrp_vm_id = hvb->vcp_id,
+        .vrp_vcpu_id = hvb->vcpu_id,
+        .vrp_init_state = {
+            .vrs_gprs[VCPU_REGS_RFLAGS] = X86_RFLAGS_INIT,
+            .vrs_gprs[VCPU_REGS_RIP] = gpa_ep,
+            .vrs_gprs[VCPU_REGS_RSP] = hv->mem_size - 8,
+            .vrs_gprs[VCPU_REGS_RDI] = X86_BOOT_INFO_BASE,
+            .vrs_crs[VCPU_REGS_CR0] = X86_CR0_INIT,
+            .vrs_crs[VCPU_REGS_CR3] = X86_CR3_INIT,
+            .vrs_crs[VCPU_REGS_CR4] = (X86_CR4_INIT | X86_CR4_VMXE),
+            .vrs_sregs[VCPU_REGS_CS] = sreg_to_vsi(&ukvm_x86_sreg_code),
+            .vrs_sregs[VCPU_REGS_DS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_ES] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_FS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_GS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_SS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_gdtr = { 0x0, X86_GDTR_LIMIT, 0x0, X86_GDT_BASE},
+            .vrs_idtr = { 0x0, 0xFFFF, 0x0, 0x0},
+            .vrs_sregs[VCPU_REGS_LDTR] = sreg_to_vsi(&ukvm_x86_sreg_unusable),
+            .vrs_sregs[VCPU_REGS_TR] = sreg_to_vsi(&ukvm_x86_sreg_tr),
+            .vrs_msrs[VCPU_REGS_EFER] = X86_EFER_INIT,
+            .vrs_msrs[VCPU_REGS_STAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_LSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_CSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_SFMASK] = 0ULL,
+            .vrs_msrs[VCPU_REGS_KGSBASE] = 0ULL,
+            .vrs_crs[VCPU_REGS_XCR0] = XCR0_X87
+        }
+    };
+
+    ukvm_x86_setup_gdt(hv->mem);
+    ukvm_x86_setup_pagetables(hv->mem, hv->mem_size);
+
+    struct ukvm_boot_info *bi =
+        (struct ukvm_boot_info *)(hv->mem + X86_BOOT_INFO_BASE);
+    bi->mem_size = hv->mem_size;
+    bi->kernel_end = gpa_kend;
+    bi->cmdline = X86_CMDLINE_BASE; 
+    bi->cpu.tsc_freq = get_tsc_freq();
+
+    if (ioctl(hvb->vmd_fd, VMM_IOC_RESETCPU, &vrp) < 0)
+        err(1, "Cannot reset VCPU - exiting.");
+
+    *cmdline = (char *)(hv->mem + X86_CMDLINE_BASE);
+}
+
+void ukvm_hv_vcpu_loop(struct ukvm_hv *hv)
+{
+    struct ukvm_hvb         *hvb = hv->b;
+    struct vm_run_params    *vrp;
+    struct passwd *pw;
+    uid_t uid;
+    gid_t gid;
+
+    vrp = malloc(sizeof(struct vm_run_params));
+    if (vrp == NULL)
+        err(1, "calloc vrp");
```

Comment:
the call was a malloc (not calloc), did you use malloc instead of calloc by mistake (just checking that you are not expected the zeroed out memory)